akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceFromPublisherSpec.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.actor.{ Actor, ActorSystem, Props }
+import akka.pattern.ask
+import akka.stream.Attributes
+import akka.testkit.TestKit
+import akka.util.Timeout
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AsyncWordSpecLike
+
+import scala.concurrent.duration._
+
+class SourceFromPublisherSpec
+    extends TestKit(ActorSystem("source-from-publisher-spec"))
+    with AsyncWordSpecLike
+    with Matchers {
+
+  private val MaxElements = 1
+  private val InitialElements = 1
+
+  import system.dispatcher
+  implicit private val timeout: Timeout = Timeout(5.second)
+  case object GetSourceMessagesCounts
+  case object IncrementSourceMessages
+  class CounterActor extends Actor {
+    private var sourceMessages = 0
+    override def receive: Receive = {
+      case IncrementSourceMessages => sourceMessages += 1
+      case GetSourceMessagesCounts => sender() ! sourceMessages
+    }
+  }
+  private val counterActor =
+    system.actorOf(Props(new CounterActor).withDispatcher("akka.test.stream-dispatcher"), s"counter-actor")
+
+  private val publisher =
+    Source
+      .fromIterator(() => Iterator.from(1))
+      .addAttributes(Attributes.inputBuffer(1, 1))
+      .takeWhile(_ <= 1000)
+      .map { i =>
+        counterActor ! IncrementSourceMessages
+        i
+      }
+      .runWith(Sink.asPublisher(fanout = false))
```

Comment:
I think you can simplify this greatly by using test probes for source and sink instead of iterator/actor.

Not 100% sure but I think something like this would do it (prefferrably inline in the test case):
```scala
 val publisher = TestPublisher.probe[Int]()

 Source.fromPublisher(publisher).addAttributes(Atrtributes.inputBuffer(1,1)).runWith(Sink.ignore)

publisher.expectRequest() should === (1L)
```

So that would make sure that the inputBuffer attribute is applied and the source publisher only asks for a single element (rather than the default 16)

Reply:
Thanks a lot @johanandren I do not know `TestPublisher.probe` exists :( 

I agree test will be way simpler than now